mm2src/crypto/src/mnemonic.rs

```diff
@@ -0,0 +1,380 @@
+use aes::cipher::{block_padding::Pkcs7, BlockDecryptMut, BlockEncryptMut, KeyIvInit};
+use aes::Aes256;
+use argon2::password_hash::{PasswordHasher, SaltString};
+use argon2::Argon2;
+use bip39::{Language, Mnemonic};
+use common::drop_mutability;
+use derive_more::Display;
+use hmac::{Hmac, Mac};
+use mm2_core::mm_ctx::MmArc;
+use mm2_err_handle::prelude::*;
+use sha2::Sha256;
+use std::convert::TryInto;
+
+const DEFAULT_WORD_COUNT: u64 = 24;
+
+type Aes256CbcEnc = cbc::Encryptor<Aes256>;
+type Aes256CbcDec = cbc::Decryptor<Aes256>;
+
+#[derive(Debug, Display, PartialEq)]
+pub enum MnemonicError {
+    #[display(fmt = "BIP39 mnemonic error: {}", _0)]
+    BIP39Error(String),
+    #[display(fmt = "Error generating random bytes: {}", _0)]
+    UnableToGenerateRandomBytes(String),
+    #[display(fmt = "Error hashing password: {}", _0)]
+    PasswordHashingFailed(String),
+    #[display(fmt = "AES cipher error: {}", _0)]
+    AESCipherError(String),
+    #[display(fmt = "Error decoding string: {}", _0)]
+    DecodeError(String),
+    #[display(fmt = "Error verifying HMAC tag: {}", _0)]
+    HMACError(String),
+    Internal(String),
+}
+
+impl From<bip39::Error> for MnemonicError {
+    fn from(e: bip39::Error) -> Self { MnemonicError::BIP39Error(e.to_string()) }
+}
+
+impl From<argon2::password_hash::Error> for MnemonicError {
+    fn from(e: argon2::password_hash::Error) -> Self { MnemonicError::PasswordHashingFailed(e.to_string()) }
+}
+
+impl From<base64::DecodeError> for MnemonicError {
+    fn from(e: base64::DecodeError) -> Self { MnemonicError::DecodeError(e.to_string()) }
+}
+
+/// Parameters for the Argon2 key derivation function.
+///
+/// This struct defines the configuration parameters used by Argon2, one of the
+/// most secure and widely used key derivation functions, especially for
+/// password hashing.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Argon2Params {
+    /// The specific variant of the Argon2 algorithm used (e.g., Argon2id).
+    algorithm: String,
+
+    /// The version of the Argon2 algorithm (e.g., 0x13 for the latest version).
+    version: String,
+
+    /// The memory cost parameter defining the memory usage of the algorithm.
+    /// Expressed in kibibytes (KiB).
+    m_cost: u32,
+
+    /// The time cost parameter defining the execution time and number of
+    /// iterations of the algorithm.
+    t_cost: u32,
+
+    /// The parallelism cost parameter defining the number of parallel threads.
+    p_cost: u32,
+}
+
+impl Default for Argon2Params {
+    fn default() -> Self {
+        Argon2Params {
+            algorithm: "Argon2id".to_string(),
+            version: "0x13".to_string(),
+            m_cost: 65536,
+            t_cost: 2,
+            p_cost: 1,
```

Comment:
It's worth using `const` or enums here for all parameters.

Reply:
done